### ISSUE TITLE
krakend cli / github access token / add custom headers

### ISSUE THREAD

**Issue by roscopecoltran** [devopsfaith/krakend#43]
Hi,

Hope you are all well !

# **1. Krakend CLI**

I was not able not find the executable/binary, mentioned in your documentation, http://www.krakend.io/docs/overview/usage/, with the output name: **krakend**. Only the set of examples (like gin) seems to be available. That's quite annoying as I wanted to try Krakend with the CLI version first. 

**Is it something available as open source ?** that would be awesome for test purposes.

# **2. Append custom headers**

**Also, is it possible to add custom headers ?** For example, I would like to get returned the **topics** property for my github search requests. It would imply to add the following header ('Accept', 'application/vnd.github.mercy-preview+json'). (ref: https://developer.github.com/v3/search/#search)

For the code search, I would be required to pass the personal token. 

**How Krakend is managing authentication to third-party providers for the moment ?** 

Is it possible to add **an example to the[ krakend-playground](https://github.com/devopsfaith/krakend-playground)** with a small group of endpoints aggregating some github search results, via the api v3, for a query submitted by POST or GET  (repo,code,commits, issues) ?

# **3. Aggregate many infos/sources for developers**

Last question, how many of these possible engines could aggregate with krakend compared to searx ? https://github.com/asciimoo/searx/blob/master/searx/settings.yml

I would like to create a **cli based utility** to aggregate dev oriented from several providers like github (starred), stackoverflow, reddit and hackernews (much more like howdoi). It would return some insights by parsing some stderr or logs, or more directly by entering a query.

refs. 
- **[ferret](https://github.com/yieldbot/ferret/tree/master/providers)** (golang) Ferret is a search engine that unifies search results from Github, Slack, Trello and more
- **[limo](https://github.com/hoop33/limo)** (golang)  a CLI for managing/searching your starred Git repositories
- **[searx](https://github.com/asciimoo/searx )** (python) hackable meta search engine
- **[searx admin](https://github.com/kvch/searx-admin)** (python), allowing to manage/edit aggregation rules

![engineslist](https://raw.githubusercontent.com/kvch/searx-admin/master/docs/images/engineslist.png)

Cheers,
Richard

**Comment by kpacha** [comment]
Hi, Richard!

Sorry for the late reply. I'm glad to see you back and playing with the krakend!

### 1. Krakend CLI

The KrakenD binary is the free build listed here: http://www.krakend.io/features/. You can download, install and use it for free without any other restrictions than the limits described there, but it is not a full open source project yet.

Check the [installation documentation](http://www.krakend.io/docs/overview/installing/) and the [download](http://www.krakend.io/download/) page for more details and see if it fits the requirements of your tests.

The open source project is the framework we created in the process of building that product. It is designed to be used as a library and the binary examples are just complementary demonstrations of different features, components and adaptors of the lib.

### 2. Append custom headers

The *EndpointHandler* of your selected router is the component you should check for that feature. The current implementations are kind of simple and they just use a private value for the list of headers to pass (see https://github.com/devopsfaith/krakend/blob/master/router/gin/endpoint.go#L76 & https://github.com/devopsfaith/krakend/blob/master/router/mux/endpoint.go#L89). I think a new wrapper will be required instead of a direct implementation into the `NewRequest` function (and variants) and, also, some implementations of the `config.ConfigGetter` [interface](https://github.com/devopsfaith/krakend/blob/master/config/config.go#L108) in order to properly build the wrapper defined in the config. I can imagine one wrapper for passing the request values of the withelisted headers and one for adding custom headers with the values defined in the config.

What kind of authentication to third-party providers are you thinking of? Out of the box, you have OAuth client credentials grant supported in the free and enterprise versions and there are tons of good libs to use if you are going for the open source based, customized build. Here you have a complete example of a JWT integration: https://github.com/devopsfaith/krakend/blob/master/examples/jwt/main.go

Regarding the github example: yes, there is an initial implementation already in the playground you can extend: https://github.com/devopsfaith/krakend-playground/blob/master/krakend/krakend.json#L11 I'm looking forward to see what do you have in mind!

### 3. Aggregate many infos/sources for developers

I'm not familiar with the searx project but, after a quick review of the `engines` [package](https://github.com/asciimoo/searx/tree/master/searx/engines) I don't see a limitation. From the top of my head, you can either:

+ Define the logic of each source in the backend section of the config file (like in the playground). If you want to reuse this backend definitions, maybe you can set a dedicated endpoint for each backend and create the 'public' endpoint that aggregates all the results from the engine-like endpoints. This is the lazy-boy solution but I'm afraid it will add to much rigidity to your project.

+ Create a custom middleware for each engine and inject them on building time. Again, the new `config.ConfigGetter` interface will help you with the required custom configurations so you can define custom configuration structs for those new proxy middlewares. This solution requires some coding but I think it won't be hard at all.

The plugglable nature of the framework supports all kind of manipulations by injecting almost any custom proxy and router component. It's open source, therefore collaborations are welcome!

So, with the examples you gave, I think krakend OS will fit your requirements perfectly. Write a basic prototype and let the community know how it goes!

Cheers

**Comment by roscopecoltran** [comment]
Hey,

Thanks for the reply !

**## Current status of experimental work:**

I already started to play and make some experiements with Krakend like pre-processing the response with wildcard rules to blacklist or whitelist some fields (arrays,...) with the mxj package.
**Ref.** 
- https://github.com/roscopecoltran/krakend/blob/sniperkit/encoding/json.go

The same for custom headers
**Ref.** 
- https://github.com/roscopecoltran/krakend/blob/sniperkit/proxy/request.go#L49-L98

Current yaml configuration file: 
https://github.com/roscopecoltran/krakend/blob/sniperkit/shared/conf.d/krakend/config-dev.yaml

Also, I started to check if I could use [qor admin](https://github.com/qor/admin) as web admin interface to manage a directory of backends and their requests. The ideal scenario would be to preload some raml or swagger files to help to setup/validate a backend config.
**Refs:**
- https://api.apis.guru/v2/list.json
- https://api.apis.guru/v2/specs/github.com/v3/swagger.yaml
- https://github.com/go-raml/raml/blob/v0/samples/github/github-api-v3.raml

## **New set of questions:**
1. How would u see to trigger some conditional sub-requests (in serie or in parallel) ? the idea would be to add some flow based sub-requests per backend or endpoint.
  - **Refs**. 
       - Mixed Request Queues: https://github.com/xh3b4sd/taskq/blob/master/taskq_mixed_test.go
       - Flow Based su-requests: https://github.com/mackristof/react-webapp/blob/master/main/react-webapp.go
  - **Use case**: would be to fetch github user info while requesting a github search and some other request like twitter API for user info.

2. How would you see to preload some backends/providers profiles and hook them in the config files with backend_id ?
  - **Use case**: Avoid redundancy of backend configuration or to auto suggest some backend configuration based on their API definitions

Have an awesome week-end and all my thoughts to Barcelona ^^

Cheers,
Richard

**Comment by kpacha** [comment]
hi!

> I already started to play and make some experiements with Krakend like pre-processing the response with wildcard rules to blacklist or whitelist some fields (arrays,...) with the mxj package

maybe you could check the #40 PR. it adds the interfaces that allow you to inject thinks like that one.

Regarding the extra questions, I'd rather open a new issue for each one, so people could find them in the future

cheers!

**Comment by kpacha** [comment]
@roscopecoltran check out the external modules listed in the [krakend-contrib](https://github.com/devopsfaith/krakend-contrib) project and the `0.4-dev` branch of the framework.

I'm pretty sure you'll find an example for almost every feature described in this issue (custom encoding components, complex request and response manipulation, pre-loaded modules, etc)
